# Text CLASSIFICATION
Text classification of news groups messages.
## DESCRIPTION



## Implementation
we will be using newsgroups dataset from sklearn.
The dataset
is a collection of messages from 20 news groups including comp.graphics,
rec.sport.hockey, sci.space, talk.religion.misc etc.

      classification method:-- Naive Bayes
      word count features:-- tf-idf
      classifier:-- Multinomial Naive Bayes classifier


following steps to are being followed in implementation:--

1. get the sample data 
2. clean the data
3. create word matrix usinf tf-idf
4. train the data
5. pass new data in the model to classify the data.

one very important aspect here is that, we keep a proportion of labelled data, to test on the trained model which will give us idea about the accuracy of "MODEL". 
usually the partition is 80% - 20%.


In [ ]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
# Get the training dataset for the specified categoires
categories = ['rec.sport.hockey', 'talk.religion.misc',
              'comp.graphics', 'sci.space']
training_data = fetch_20newsgroups(subset='train', categories=categories)

In [ ]:
# Create the tf-idf transformer
tfidf = TfidfVectorizer(use_idf=True)
# tfidf = TfidfVectorizer(use_idf=False)
training_tfidf = tfidf.fit_transform(training_data.data)
print(training_tfidf.shape)

# Train a Multinomial Naive Bayes classifier
classifier = MultinomialNB().fit(training_tfidf, training_data.target)


(2154, 35956)


In [ ]:
# Create the tf-idf transformer

tfidf = TfidfVectorizer(use_idf=False)
training_tfidf = tfidf.fit_transform(training_data.data)
print(training_tfidf.shape)

#Train a Multinomial Naive Bayes classifier
classifier = MultinomialNB().fit(training_tfidf, training_data.target)

In [ ]:
from sklearn import metrics

testing_data = fetch_20newsgroups(subset='test', categories=categories)
testing_tfidf = tfidf.transform(testing_data.data)
predictions = classifier.predict(testing_tfidf)
print(metrics.classification_report(testing_data.target, predictions, target_names=categories))


                    precision    recall  f1-score   support

  rec.sport.hockey       0.97      0.90      0.93       389
talk.religion.misc       0.93      0.99      0.96       399
     comp.graphics       0.83      0.98      0.90       394
         sci.space       1.00      0.72      0.84       251

          accuracy                           0.92      1433
         macro avg       0.93      0.90      0.91      1433
      weighted avg       0.93      0.92      0.92      1433



In [ ]:
errors = [i for i in range(len(predictions)) if predictions[i] != testing_data.target[i]]

for i, post_id in enumerate(errors[:5]):
  print("------------------------------------------------------------------")
  print("%s --> %s\n" %(testing_data.target_names[testing_data.target[post_id]], 
                      testing_data.target_names[predictions[post_id]]))
  print(testing_data.data[post_id])


------------------------------------------------------------------
comp.graphics --> sci.space

From: robert@slipknot.rain.com (Robert Reed)
Subject: Re: ACM SIGGRAPH (and ACM in general)
Reply-To: Robert Reed <robert@slipknot.rain.com>
Organization: Home Animation Ltd.
Lines: 50

In article <1993Apr29.023508.11556@koko.csustan.edu> rsc@altair.csustan.edu (Steve Cunningham) writes:
|
|And no, SIGGRAPH 93 has not skipped town -- we're preparing the best
|SIGGRAPH conference yet!

Speaking of SIGGRAPH, I just went through the ordeal of my annual registration
for SIGGRAPH and re-upping of membership in the ACM last night, and was I ever
grossed out!  The new prices for membership are almost highway robbery!

For example:

	SIGGRAPH basic fee went from $26 last year to $59 this year for the same
	thing, a 127% increase.  Those facile enough to arrange a trip to the
	annual conference could reduce this to $27 by selecting SIGGRAPH Lite,
	which means SIGGRAPH is charging an additional $32 (o

## Discussion

Using Naive Bayes multinomial classifier we could train the model to classify the text data, for multi level class.

* we explored both approaches with tfidf and without tfidf.
      tfidf = TfidfVectorizer(use_idf=False)

To know the model metric, we have used here matric package from sklearn library.


    * below is model summary with its precision , recall and
     f1 score.
    * model summary

                        precision    recall  f1-score   


    rec.sport.hockey      0.97      0.90      0.93       
    talk.religion.misc    0.93      0.99      0.96      
    comp.graphics         0.83      0.98      0.90       
    sci.space            1.00      0.72       0.84      

    accuracy                                  0.92       


    
    * Here F1 score is most important, as it is combination of both precision and recall.

    * f1 score helps balance the two metrics.



* Errors

     * we have also explore the errors , by printing the wrong predictions made by our model.

     * eg:  comp.graphics --> sci.space
       here actual class is comp.graphics
       but predicted one is sci.space